apps/etl-func/src/adapters/functions/messages-ingestion.ts

```diff
@@ -16,23 +12,20 @@ const messagesIngestionHandler =
     telemetryService: TelemetryService,
   ): CosmosDBHandler =>
   async (documents: unknown[], context: InvocationContext) => {
-    const parsedMessagesMetadataOrZodError = documents.map((input) => {
-      const parsedInput = messageMetadataSchema.safeParse(input);
-      if (parsedInput.success) return parsedInput.data;
-      else return parsedInput.error;
-    });
-
     // filter the array in order to get only valid MessageMetadata with
     // isPending = false
-    const nonPendingMessageMetadata = parsedMessagesMetadataOrZodError.filter(
-      (document): document is MessageMetadata =>
-        !(document instanceof ZodError) && !document.isPending,
-    );
+    const nonPendingMessageMetadata = documents
+      .filter((document) => {
+        const parsedInput = messageMetadataSchema.safeParse(document);
+        return parsedInput.success && !parsedInput.data.isPending;
+      })
+      .map((document) => messageMetadataSchema.parse(document));
 
     // get all malformed documents so we can send them to the error repository
-    const malformedDocuments = parsedMessagesMetadataOrZodError.filter(
-      (input): input is ZodError => input instanceof ZodError,
-    );
+    const malformedDocuments = documents.filter((document) => {
+      const parsedInput = messageMetadataSchema.safeParse(document);
+      return !parsedInput.success;
+    });
```

Comment:
If you do a map before the filter as I wrote in the comment before, you can directly filter with a guard instead of calling the safeParse again.